academy/ibc-dev/ibcmw-integrate.md

```diff
@@ -0,0 +1,156 @@
+# Integrating IBC middleware into a chain
+
+<HighlightBox type="learning">
+
+Learn how to integrate IBC middleware(s) with a base application to your chain. The following document only applies for Cosmos SDK chains.
+
+</HighlightBox>
+
+We've now seen how to develop an IBC middleware, the final step in order to use it is to integrate it into the chain. We can distinguish between a single piece of middleware or a stack of middleware and between middleware that is stateful or stateless.
+
+To integrate our middleware (stack) we must do the following in `app.go`:
+
+- If the middleware is maintaining its own state and/or processing SDK messages, then it should create and register its SDK module **only once** with the module manager
+
+    ```go
+    //app.go
+
+    // create a keeper for the stateful middleware
+    mw1Keeper := mw1.NewKeeper(storeKey1)
+    mw3Keeper := mw3.NewKeeper(storeKey3)
+
+    // instantiate the middleware as IBCModules
+    mw1IBCModule := mw1.NewIBCModule(mw1Keeper)
+    mw2IBCModule := mw2.NewIBCModule() // middleware2 is stateless middleware
+    mw3IBCModule := mw3.NewIBCModule(mw3Keeper)
+
+    // register middleware in app module
+    // if the module maintains independent state and/or processes sdk.Msgs
+    app.moduleManager = module.NewManager(
+        ...
+        mw1.NewAppModule(mw1Keeper),
+        mw3.NewAppModule(mw3Keeper),
+        transfer.NewAppModule(transferKeeper),
+        custom.NewAppModule(customKeeper)
+    )
+    
+    ```
+
+    <ExpansionPanel title="Example code">
+    <!-- TODO: put code above in the expansion panel -->
+    </ExpansionPanel>
+
+- As mentioned in the previous section, in the case where the middleware wishes to send a packet or acknowledgment without the involvement of the underlying application, it should be given access to the same `scopedKeeper` as the base application so that it can retrieve the capabilities by itself.
```

Comment:
Should we provide example code how to actually do this? Right now we define `scopedKeepers` but not using them, which looks somewhat confusing...

Reply:
Yea probably have at least one middleware stack where this is done. The rest do not need their capabilityKeepers defined in the pseudocode

Reply:
We can explain that they are wrapped in the base application's keeper